File: analytics.py
"""
analytics.py
============
Framework-free statistical functions computed over a single health metric's
time series: baselines, anomaly detection, trend, period comparison, and
cross-metric correlation.

Deliberately dependency-free (standard library only, mirroring logic.py) so
these are unit-testable without fastmcp and reusable directly from
server.py's Layer-2/Layer-3 MCP tools. None of this touches the database —
every function takes an already-fetched (date, value) series. That keeps
this module with zero opinions about SQLite, connections, or which metrics
exist; server.py stays responsible for reading rows (via
logic.readonly_connection) and for enforcing HEALTH_PRIVATE_FIELDS before
anything here ever sees a value.
"""

from __future__ import annotations

import statistics
from datetime import date
from typing import NamedTuple


class Point(NamedTuple):
    day: date
    value: float


def _values(series: list[Point]) -> list[float]:
    return [p.value for p in series]


def baseline(series: list[Point]) -> dict[str, float | None]:
    """Mean, median, population stdev, and sample size for a series.

    Returns Nones (n=0) for an empty series rather than raising — "no data
    logged yet for this metric" is an expected, tool-reportable case for a
    Layer-3 caller, not an exceptional one.
    """
    values = _values(series)
    n = len(values)
    if n == 0:
        return {"mean": None, "median": None, "stdev": None, "n": 0}
    return {
        "mean": round(statistics.fmean(values), 2),
        "median": round(statistics.median(values), 2),
        "stdev": round(statistics.pstdev(values), 2) if n > 1 else 0.0,
        "n": n,
    }


def detect_anomalies(series: list[Point], threshold: float = 3.5) -> list[dict]:
    """Flag points that deviate sharply from the series' own baseline,
    using a modified z-score built on median + MAD (median absolute
    deviation) rather than mean/stdev.

    MAD is robust to the very outliers it's trying to detect, and to the
    small, noisy samples typical of personal health data (30-90 days),
    where a couple of bad nights would otherwise drag the mean/stdev
    enough to mask themselves. threshold=3.5 is Iglewicz & Hoaglin's
    standard cutoff for "this point doesn't belong."
    """
    if len(series) < 5:
        return []  # not enough history to call anything an "anomaly"
    values = _values(series)
    med = statistics.median(values)
    abs_devs = [abs(v - med) for v in values]
    mad = statistics.median(abs_devs)
    if mad == 0:
        return []  # a flat series has no meaningful spread to deviate from
    anomalies = []
    for point, dev in zip(series, abs_devs, strict=True):
        modified_z = 0.6745 * dev / mad
        if modified_z >= threshold:
            anomalies.append(
                {
                    "date": point.day.isoformat(),
                    "value": point.value,
                    "modified_z_score": round(modified_z, 2),
                    "direction": "above" if point.value > med else "below",
                }
            )
    return anomalies


def calculate_trend(series: list[Point]) -> dict:
    """Direction and slope of an ordinary-least-squares fit against actual
    calendar time (days elapsed since the first observation), not position
    in the list.

    A missing day is a gap, not a step of equal size to a logged one — a
    reading on Jan 1 followed by the next on Jan 20 is 19 days apart, not
    "the next point." Fitting against position instead of calendar time
    would silently compress that 19-day gap to a single step, distorting
    slope_per_day without the caller ever finding out. Points do not need
    to be evenly spaced or gap-free for this to be correct: OLS against
    x = elapsed_days handles irregular spacing natively. Coverage/gap
    reporting (evidence.py) is a separate, complementary signal about how
    much of the window is backed by data — it should never be doing the
    job of fixing a distorted time axis here.
    """
    n = len(series)
    if n < 3:
        return {
            "direction": "insufficient_data",
            "slope_per_day": None,
            "r_squared": None,
            "n": n,
            "span_days": None,
        }
    origin = series[0].day
    xs = [(p.day - origin).days for p in series]
    ys = _values(series)
    span_days = xs[-1] - xs[0]
    x_mean = sum(xs) / n
    y_mean = sum(ys) / n
    ss_xy = sum((x - x_mean) * (y - y_mean) for x, y in zip(xs, ys, strict=True))
    ss_xx = sum((x - x_mean) ** 2 for x in xs)
    if ss_xx == 0:
        # Every point falls on the same calendar day (e.g. duplicate-day
        # rows) — there's no time axis to regress against.
        return {"direction": "flat", "slope_per_day": 0.0, "r_squared": 0.0, "n": n, "span_days": span_days}
    slope = ss_xy / ss_xx
    intercept = y_mean - slope * x_mean
    ss_tot = sum((y - y_mean) ** 2 for y in ys)
    ss_res = sum((y - (slope * x + intercept)) ** 2 for x, y in zip(xs, ys, strict=True))
    r_squared = 1.0 if ss_tot == 0 else 1 - ss_res / ss_tot
    direction = "flat" if abs(slope) < 1e-9 else ("increasing" if slope > 0 else "decreasing")
    return {
        "direction": direction,
        "slope_per_day": round(slope, 4),
        "r_squared": round(r_squared, 3),
        "n": n,
        "span_days": span_days,
    }


def compare_periods(series_a: list[Point], series_b: list[Point]) -> dict:
    """Compare two (typically non-overlapping) periods of the same metric.

    series_a is treated as the "current"/later period, series_b as the
    baseline it's measured against (e.g. this week vs. the prior 4 weeks).
    """
    base_a, base_b = baseline(series_a), baseline(series_b)
    if base_a["mean"] is None or base_b["mean"] is None:
        return {"period_a": base_a, "period_b": base_b, "delta": None, "pct_change": None}
    delta = round(base_a["mean"] - base_b["mean"], 2)
    pct_change = round((delta / base_b["mean"]) * 100, 1) if base_b["mean"] != 0 else None
    return {"period_a": base_a, "period_b": base_b, "delta": delta, "pct_change": pct_change}


def _sample_confidence(n: int) -> str:
    """Bucket an overlap/sample count into a coarse adequacy label for
    correlation results — product-level guardrails, not a universal
    statistical threshold. Deliberately silent about the correlation's
    magnitude or significance; this only says how much paired data backed
    it. See find_correlations.
    """
    if n < 4:
        return "insufficient"
    if n < 10:
        return "very_limited"
    if n < 20:
        return "limited"
    if n < 30:
        return "moderate"
    return "strong_sample"


def find_correlations(series_a: list[Point], series_b: list[Point], lag_days: int = 0) -> dict:
    """Pearson correlation between two metrics' series, joined by date.

    lag_days > 0 shifts series_b backward before joining — lag_days=1 tests
    whether series_a on day N predicts series_b on day N+1 (e.g. "does poor
    sleep tonight predict lower steps tomorrow?"). Only dates present in
    both series after the shift are used.

    Every return path includes "sample_confidence" (see _sample_confidence)
    — a bucketed read on the overlap count "n" alone, not on the strength
    of "r" — so a caller can't mistake a high r from a thin sample for a
    settled relationship.
    """
    shifted_b = {p.day.toordinal() - lag_days: p.value for p in series_b}
    pairs = [(p.value, shifted_b[p.day.toordinal()]) for p in series_a if p.day.toordinal() in shifted_b]
    n = len(pairs)
    if n < 4:
        return {
            "r": None,
            "n": n,
            "lag_days": lag_days,
            "note": "Not enough overlapping days to compute a correlation.",
            "sample_confidence": _sample_confidence(n),
        }
    xs, ys = zip(*pairs, strict=True)
    try:
        r = statistics.correlation(xs, ys)
    except statistics.StatisticsError:
        return {
            "r": None,
            "n": n,
            "lag_days": lag_days,
            "note": "No variance in one of the two series.",
            "sample_confidence": _sample_confidence(n),
        }
    return {"r": round(r, 3), "n": n, "lag_days": lag_days, "sample_confidence": _sample_confidence(n)}

File: node/src/analytics.ts
export interface Point {
  day: Date; // UTC midnight
  value: number;
}

function values(series: Point[]): number[] {
  return series.map((p) => p.value);
}

function mean(xs: number[]): number {
  return xs.reduce((a, b) => a + b, 0) / xs.length;
}

function median(xs: number[]): number {
  const sorted = [...xs].sort((a, b) => a - b);
  const mid = Math.floor(sorted.length / 2);
  return sorted.length % 2 !== 0 ? sorted[mid] : (sorted[mid - 1] + sorted[mid]) / 2;
}

function pstdev(xs: number[]): number {
  const m = mean(xs);
  return Math.sqrt(mean(xs.map((x) => (x - m) ** 2)));
}

/** Population Pearson correlation, matching Python's statistics.correlation. */
function correlation(xs: number[], ys: number[]): number {
  const n = xs.length;
  const mx = mean(xs);
  const my = mean(ys);
  let num = 0,
    dx2 = 0,
    dy2 = 0;
  for (let i = 0; i < n; i++) {
    const dx = xs[i] - mx;
    const dy = ys[i] - my;
    num += dx * dy;
    dx2 += dx * dx;
    dy2 += dy * dy;
  }
  if (dx2 === 0 || dy2 === 0) throw new Error("No variance in one of the two series.");
  return num / Math.sqrt(dx2 * dy2);
}

export interface BaselineStats {
  mean: number | null;
  median: number | null;
  stdev: number | null;
  n: number;
}

/** Mean, median, population stdev, and sample size for a series. */
export function baseline(series: Point[]): BaselineStats {
  const vals = values(series);
  const n = vals.length;
  if (n === 0) return { mean: null, median: null, stdev: null, n: 0 };
  return {
    mean: Math.round(mean(vals) * 100) / 100,
    median: Math.round(median(vals) * 100) / 100,
    stdev: n > 1 ? Math.round(pstdev(vals) * 100) / 100 : 0.0,
    n,
  };
}

export interface Anomaly {
  date: string;
  value: number;
  modified_z_score: number;
  direction: "above" | "below";
}

/**
 * Flag points that deviate sharply from the series' own baseline using a
 * modified z-score built on median + MAD (median absolute deviation).
 * threshold=3.5 is Iglewicz & Hoaglin's standard cutoff.
 */
export function detectAnomalies(series: Point[], threshold = 3.5): Anomaly[] {
  if (series.length < 5) return [];
  const vals = values(series);
  const med = median(vals);
  const absDevs = vals.map((v) => Math.abs(v - med));
  const mad = median(absDevs);
  if (mad === 0) return [];
  const anomalies: Anomaly[] = [];
  series.forEach((point, i) => {
    const modifiedZ = (0.6745 * absDevs[i]) / mad;
    if (modifiedZ >= threshold) {
      anomalies.push({
        date: point.day.toISOString().slice(0, 10),
        value: point.value,
        modified_z_score: Math.round(modifiedZ * 100) / 100,
        direction: point.value > med ? "above" : "below",
      });
    }
  });
  return anomalies;
}

export interface TrendStats {
  direction: "insufficient_data" | "flat" | "increasing" | "decreasing";
  slope_per_day: number | null;
  r_squared: number | null;
  n: number;
}

/** Direction and slope of an OLS fit against day index (0, 1, 2, ...). */
export function calculateTrend(series: Point[]): TrendStats {
  const n = series.length;
  if (n < 3) return { direction: "insufficient_data", slope_per_day: null, r_squared: null, n };
  const xs = Array.from({ length: n }, (_, i) => i);
  const ys = values(series);
  const xMean = mean(xs);
  const yMean = mean(ys);
  let ssXy = 0,
    ssXx = 0;
  for (let i = 0; i < n; i++) {
    ssXy += (xs[i] - xMean) * (ys[i] - yMean);
    ssXx += (xs[i] - xMean) ** 2;
  }
  if (ssXx === 0) return { direction: "flat", slope_per_day: 0.0, r_squared: 0.0, n };
  const slope = ssXy / ssXx;
  const intercept = yMean - slope * xMean;
  const ssTot = ys.reduce((a, y) => a + (y - yMean) ** 2, 0);
  const ssRes = xs.reduce((a, x, i) => a + (ys[i] - (slope * x + intercept)) ** 2, 0);
  const rSquared = ssTot === 0 ? 1.0 : 1 - ssRes / ssTot;
  const direction = Math.abs(slope) < 1e-9 ? "flat" : slope > 0 ? "increasing" : "decreasing";
  return {
    direction,
    slope_per_day: Math.round(slope * 10000) / 10000,
    r_squared: Math.round(rSquared * 1000) / 1000,
    n,
  };
}

export interface ComparePeriodsResult {
  period_a: BaselineStats;
  period_b: BaselineStats;
  delta: number | null;
  pct_change: number | null;
}

/** Compare two (typically non-overlapping) periods of the same metric. */
export function comparePeriods(seriesA: Point[], seriesB: Point[]): ComparePeriodsResult {
  const baseA = baseline(seriesA);
  const baseB = baseline(seriesB);
  if (baseA.mean === null || baseB.mean === null) {
    return { period_a: baseA, period_b: baseB, delta: null, pct_change: null };
  }
  const delta = Math.round((baseA.mean - baseB.mean) * 100) / 100;
  const pctChange = baseB.mean !== 0 ? Math.round((delta / baseB.mean) * 100 * 10) / 10 : null;
  return { period_a: baseA, period_b: baseB, delta, pct_change: pctChange };
}

export interface CorrelationResult {
  r: number | null;
  n: number;
  lag_days: number;
  note?: string;
}

/**
 * Pearson correlation between two metrics' series, joined by date.
 * lag_days > 0 shifts series_b backward before joining.
 */
export function findCorrelations(seriesA: Point[], seriesB: Point[], lagDays = 0): CorrelationResult {
  const shiftedB = new Map<number, number>();
  for (const p of seriesB) {
    shiftedB.set(Math.floor(p.day.getTime() / 86_400_000) - lagDays, p.value);
  }
  const pairs: [number, number][] = [];
  for (const p of seriesA) {
    const ord = Math.floor(p.day.getTime() / 86_400_000);
    if (shiftedB.has(ord)) pairs.push([p.value, shiftedB.get(ord)!]);
  }
  const n = pairs.length;
  if (n < 4) {
    return { r: null, n, lag_days: lagDays, note: "Not enough overlapping days to compute a correlation." };
  }
  const xs = pairs.map((p) => p[0]);
  const ys = pairs.map((p) => p[1]);
  try {
    const r = correlation(xs, ys);
    return { r: Math.round(r * 1000) / 1000, n, lag_days: lagDays };
  } catch {
    return { r: null, n, lag_days: lagDays, note: "No variance in one of the two series." };
  }
}
